### ISSUE TITLE
Initiator should not remove payment task if the block of the lock expiry comes and there has been a refund transfer

### ISSUE THREAD

**Issue by LefterisJP** [raiden-network/raiden#3183]
## Problem Definition

This is yet another episode on the "Hanging Transfers/Nonce mismatch" saga. The logs were provided by @kelsos [here](https://drive.google.com/file/d/1Itmhsnm1pHdifMQJucRVXn-M1DnF7V-L/view?usp=sharing).

The relevant nodes are `0x82641569b2062B545431cF6D7F0A418582865ba7` who experiences the nonce mismatch and his partner `0x504300c525cbe91adb3fe0944fe1f56f5162c75c`. The setup and topology is exactly the same as @kelsos has described [here](https://github.com/raiden-network/raiden/issues/3146#issuecomment-446936518).

### Sudying the logs and DBs

Studying the logs the situations seems identical to the previous hanging transfer up to point described [here](https://github.com/raiden-network/raiden/issues/3146#issuecomment-447027876).

The first mismatch is seen here

```
03282 State change	2018-12-17 15:08:17.348691	raiden.raiden_service	debug	greenlet_name=sync_handle_response-s1134736_3554614_0_1_1155_1_1_7686_1 node=82641569 state_change={"secret": "0x3cf829a0c7d6f7f58e847cb844d691305a362174d9123f047caa50a2fc722d1b", "routes": [], "transfer": {"message_identifier": "3803582215950841833", "payment_identifier": "1545059296", "token": "0x28104EE15e1c70c421150865C3fb731c426929E6", "balance_proof": {"nonce": 7, "transferred_amount": "0", "locked_amount": "1", "locksroot": "0x7641a29976ebb579deff64e7421e797429c66c868af45a3ca77b4a267ce2997e", "token_network_identifier": "0x15008736b8eDf6EaB458D18440876aBd65e57837", "channel_identifier": "3", "message_hash": "0xe9d64a554544bc365a27bb210c60f374d7093a04fbcc392dc53b6ca3722ab4fd", "signature": "0x525d95a48bb8615f626fa71cc31283e4a08036b16fe5cadbb852034172d3203637c6ca26b5a6e49c6c6c51ca85c7db0d3acef6ca35afb02203128e7549f51b441c", "sender": "0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C", "chain_id": 17, "balance_hash": "0xfac4e0821f9f5625ff3b8fd3855288acbc4f0bb4b68d6a595a4787a0c06df199", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "lock": {"amount": 1, "expiration": 6863, "secrethash": "0x1d59b33b27d4dbaa57a0d545bb412aa6739108590610383325d02fc71b8b8343", "_type": "raiden.transfer.state.HashTimeLockState", "_version": 0}, "initiator": "0x82641569b2062B545431cF6D7F0A418582865ba7", "target": "0x60FFe4daDE358246C49C4A16bB5b430d333B5Ce9", "_type": "raiden.transfer.mediated_transfer.state.LockedTransferSignedState", "_version": 0}, "balance_proof": {"nonce": 7, "transferred_amount": "0", "locked_amount": "1", "locksroot": "0x7641a29976ebb579deff64e7421e797429c66c868af45a3ca77b4a267ce2997e", "token_network_identifier": "0x15008736b8eDf6EaB458D18440876aBd65e57837", "channel_identifier": "3", "message_hash": "0xe9d64a554544bc365a27bb210c60f374d7093a04fbcc392dc53b6ca3722ab4fd", "signature": "0x525d95a48bb8615f626fa71cc31283e4a08036b16fe5cadbb852034172d3203637c6ca26b5a6e49c6c6c51ca85c7db0d3acef6ca35afb02203128e7549f51b441c", "sender": "0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C", "chain_id": 17, "balance_hash": "0xfac4e0821f9f5625ff3b8fd3855288acbc4f0bb4b68d6a595a4787a0c06df199", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "_type": "raiden.transfer.mediated_transfer.state_change.ReceiveTransferRefundCancelRoute", "_version": 0}
03283 Raiden event	2018-12-17 15:08:17.401663	raiden.raiden_service	debug	greenlet_name=sync_handle_response-s1134736_3554614_0_1_1155_1_1_7686_1 node=82641569 raiden_event={"payment_identifier": "1545059296", "reason": "Invalid LockedTransfer message. Nonce did not change sequentially, expected: 4 got: 7.", "_type": "raiden.transfer.events.EventInvalidReceivedLockedTransfer", "_version": 0}
```

- channel_identifier: 3
- sender: 0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C
- expected nonce: 4  got: 7

Going back to find the message with nonce `4` we see:

```
00332 Incoming messages	2018-12-17 15:06:02.284422	raiden.network.transport.matrix	debug	greenlet_name=sync_handle_response-s1133922_3554565_0_1_1155_1_1_7685_1 node=82641569 current_user=@0x82641569b2062b545431cf6d7f0a418582865ba7:persephone.raidentransport.digitalvirtues.com sender=504300c5 messages=['<LockExpired [chainid:17 token_network_address:15008736 msg_id:1917831621134351790 channel_identifier:3 nonce:4 transferred_amount:0 locked_amount:2 locksroot:8698151e secrethash:925f9fc3]>'] room=<Room id="!jGchuqeUYjhZcnIOEe:raidentransport.ki-decentralized.com" aliases=[]> sender_user=<User id="@0x504300c525cbe91adb3fe0944fe1f56f5162c75c:persephone.raidentransport.digitalvirtues.com">
00333 Message received	2018-12-17 15:06:02.285135	raiden.network.transport.matrix	debug	greenlet_name=sync_handle_response-s1133922_3554565_0_1_1155_1_1_7685_1 node=82641569 current_user=@0x82641569b2062b545431cf6d7f0a418582865ba7:persephone.raidentransport.digitalvirtues.com message=<LockExpired [chainid:17 token_network_address:15008736 msg_id:1917831621134351790 channel_identifier:3 nonce:4 transferred_amount:0 locked_amount:2 locksroot:8698151e secrethash:925f9fc3]> sender=504300c5
00334 State change	2018-12-17 15:06:02.287489	raiden.raiden_service	debug	greenlet_name=sync_handle_response-s1133922_3554565_0_1_1155_1_1_7685_1 node=82641569 state_change={"balance_proof": {"nonce": 4, "transferred_amount": "0", "locked_amount": "2", "locksroot": "0x8698151e212f196720ca0ea6aaf0d6b04a0405ba3baaad361d31df9082c55570", "token_network_identifier": "0x15008736b8eDf6EaB458D18440876aBd65e57837", "channel_identifier": "3", "message_hash": "0x189800aa6dd49212b5c664af202038381b0f43da508f21652f827387b683b680", "signature": "0xdfdfd488e3426d744eac6dfe3f7798587e377254e6c12ec30274b75728dc534b5d96ded529b071a2d6adfcc8ab53f4f77fe353d31f3a068728780ead3d5d07851c", "sender": "0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C", "chain_id": 17, "balance_hash": "0x6a9d34913670cdf5c909cdd5903a5f6c12ae64b017d633ddce586656c6854a5f", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "secrethash": "0x925f9fc3a5115b9f4c40cb30fce54f3eb926d82a6f358c12efc41773bd3af3cf", "message_identifier": "1917831621134351790", "_type": "raiden.transfer.mediated_transfer.state_change.ReceiveLockExpired", "_version": 0}
```

Which is a lock expired message. Its secrethash is: `b'\x92_\x9f\xc3\xa5\x11[\x9fL@\xcb0\xfc\xe5O>\xb9&\xd8*o5\x8c\x12\xef\xc4\x17s\xbd:\xf3\xcf'`

It generates no events because there is no payment task for this secrethash just like in all the previous incarnations of the problem.

Nonce `3` for that channel is:

```
<ReceiveTransferRefundCancelRoute sender:504300c5 transfer:<LockedTransferSignedState msgid:6014380788349923296 id:1545058903 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:6484 secrethash:3a46edfb> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
```

We receive the refund, it's valid and we send a processed. But when we try to find new routes there is no new routes and we fail the transfer but keep the payment task around thanks to the work done in [one](https://github.com/raiden-network/raiden/pull/3167) of the previous PR addressing the nonce mismatches. So that couldn't be the problem.

Searching the DB a bit deeper we can find out which state change removes the `secrethash` `b'\x92_\x9f\xc3\xa5\x11[\x9fL@\xcb0\xfc\xe5O>\xb9&\xd8*o5\x8c\x12\xef\xc4\x17s\xbd:\xf3\xcf'` from the payment mapping. Turns out it's:

`<Block number=6435 gas_limit=8000000 block_hash=9894f941>`

And indeed ..

https://github.com/raiden-network/raiden/blob/7a3850f05c6591921ef6a4003d5ed574f2efebd2/raiden/transfer/mediated_transfer/initiator.py#L91-L116

It seems that the code removes the initiator payment task if we receive the block for which the lock is considered expired.

### Thoughts on the problem

This functionality should be fine for most cases since the initiator is the one who sends `LockExpired` so he does not need to wait for a `LockExpired` to arrive before he deletes the payment task. Normally nobody needs to send a LockExpired to the initiator.

Things get complicated when there is a refund transfer. Since refund transfers use the same secrethash as the original transfer and the roles are reversed, the initiator can receive a lock expired message from the partner who sent him the refund transfer even though he is actually the original initiator.

## Task

Got some WIP solution in a branch but perhaps gotta think about the proper way to address this a bit more.

**Comment by rakanalh** [comment]
Something is wrong here. Upon receiving `LockExpired` in the mediator / target as a receiver, this receiver is supposed to remove the lock from their end and reply with `Processed`. Seems like this did not happen and instead chose to send a RefundTransfer. Which means that:
1. Either the intended receiver did not receive the message altogether.
2. The message was indeed received and `SendProcessed` was queued but we sent out the refund JUST before we got the the `LockExpired` message.

In (1), we have a problem that we could confirm looking at the logs of the receiver and see if that `LockExpired` message was received/processed.
In (2), it's not a problem because the receiver would have sent out a refund, sender sends out the lock expired message and the task gets cleared on both ends. So the initiator is right to detect the refund transfer as invalid and the receiver (be it a mediator or a target) would have received the lock expired and removed the locks.

**Comment by rakanalh** [comment]
Looking at the description for this issue, i noticed the following:

1. For nonce: 3, we have an initiator task which received a refund transfer. The secret hash is: `3a46edfb`
```
<ReceiveTransferRefundCancelRoute sender:504300c5 transfer:<LockedTransferSignedState msgid:6014380788349923296 id:1545058903 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:6484 secrethash:3a46edfb> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
```

2. For nonce: 4, we have a task which received a lockexpired message with a secret hash:
`0x925f9fc3a5115b9f4c40cb30fce54f3eb926d82a6f358c12efc41773bd3af3cf`

```
state_change={"balance_proof": {"nonce": 4, "transferred_amount": "0", "locked_amount": "2", "locksroot": "0x8698151e212f196720ca0ea6aaf0d6b04a0405ba3baaad361d31df9082c55570", "token_network_identifier": "0x15008736b8eDf6EaB458D18440876aBd65e57837", "channel_identifier": "3", "message_hash": "0x189800aa6dd49212b5c664af202038381b0f43da508f21652f827387b683b680", "signature": "0xdfdfd488e3426d744eac6dfe3f7798587e377254e6c12ec30274b75728dc534b5d96ded529b071a2d6adfcc8ab53f4f77fe353d31f3a068728780ead3d5d07851c", "sender": "0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C", "chain_id": 17, "balance_hash": "0x6a9d34913670cdf5c909cdd5903a5f6c12ae64b017d633ddce586656c6854a5f", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "secrethash": "0x925f9fc3a5115b9f4c40cb30fce54f3eb926d82a6f358c12efc41773bd3af3cf", "message_identifier": "1917831621134351790", "_type": "raiden.transfer.mediated_transfer.state_change.ReceiveLockExpired", "_version": 0}
```

3. And the rejected message of nonce: 7, we received a refund transfer in an intiator task with secrethash:
`0x1d59b33b27d4dbaa57a0d545bb412aa6739108590610383325d02fc71b8b8343`
```
state_change={"secret": "0x3cf829a0c7d6f7f58e847cb844d691305a362174d9123f047caa50a2fc722d1b", "routes": [], "transfer": {"message_identifier": "3803582215950841833", "payment_identifier": "1545059296", "token": "0x28104EE15e1c70c421150865C3fb731c426929E6", "balance_proof": {"nonce": 7, "transferred_amount": "0", "locked_amount": "1", "locksroot": "0x7641a29976ebb579deff64e7421e797429c66c868af45a3ca77b4a267ce2997e", "token_network_identifier": "0x15008736b8eDf6EaB458D18440876aBd65e57837", "channel_identifier": "3", "message_hash": "0xe9d64a554544bc365a27bb210c60f374d7093a04fbcc392dc53b6ca3722ab4fd", "signature": "0x525d95a48bb8615f626fa71cc31283e4a08036b16fe5cadbb852034172d3203637c6ca26b5a6e49c6c6c51ca85c7db0d3acef6ca35afb02203128e7549f51b441c", "sender": "0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C", "chain_id": 17, "balance_hash": "0xfac4e0821f9f5625ff3b8fd3855288acbc4f0bb4b68d6a595a4787a0c06df199", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "lock": {"amount": 1, "expiration": 6863, "secrethash": "0x1d59b33b27d4dbaa57a0d545bb412aa6739108590610383325d02fc71b8b8343", "_type": "raiden.transfer.state.HashTimeLockState", "_version": 0}, "initiator": "0x82641569b2062B545431cF6D7F0A418582865ba7", "target": "0x60FFe4daDE358246C49C4A16bB5b430d333B5Ce9", "_type": "raiden.transfer.mediated_transfer.state.LockedTransferSignedState", "_version": 0}, "balance_proof": {"nonce": 7, "transferred_amount": "0", "locked_amount": "1", "locksroot": "0x7641a29976ebb579deff64e7421e797429c66c868af45a3ca77b4a267ce2997e", "token_network_identifier": "0x15008736b8eDf6EaB458D18440876aBd65e57837", "channel_identifier": "3", "message_hash": "0xe9d64a554544bc365a27bb210c60f374d7093a04fbcc392dc53b6ca3722ab4fd", "signature": "0x525d95a48bb8615f626fa71cc31283e4a08036b16fe5cadbb852034172d3203637c6ca26b5a6e49c6c6c51ca85c7db0d3acef6ca35afb02203128e7549f51b441c", "sender": "0x504300C525CbE91Adb3FE0944Fe1f56f5162C75C", "chain_id": 17, "balance_hash": "0xfac4e0821f9f5625ff3b8fd3855288acbc4f0bb4b68d6a595a4787a0c06df199", "_type": "raiden.transfer.state.BalanceProofSignedState", "_version": 0}, "_type": "raiden.transfer.mediated_transfer.state_change.ReceiveTransferRefundCancelRoute", "_version": 0}
```

From this, we know that we're talking about the same channel but 3 different tasks.

Investigating further.

**Comment by LefterisJP** [comment]
As discussed with @rakanalh in the standup also trying to see why the partner sent us nonce `7` while we expected `4` is a good question. @rakanalh will investigate why by looking at both DBs and if he finds another issue he will open a new issue and handle it there.

@LefterisJP will handle the current issue described in the first post by creating a special case of handling `LockExpired` in the initiator for refund transfers. Also by not deleting the payment task of the initiator if the block of lock expiration comes and there have been refund transfers. Because you need to wait for a lock expiration.

**Comment by rakanalh** [comment]
TL;DR: The initiator should implement handling of `LockExpired` messages to be able to clear up `channel_state.partner_state.secrethashes_to_lockedlocks` in the case the initiator receives a `RefundTransfer` whose lock expires.

Longer version:
We investigated the logs and found that for refund transfers, the mediator would send them back to the previous hop, which is the initiator in an A (initiator) -> B (Mediator) -> C (Target) topology. 
What was curious as to why there's a > 1 difference in nonces between the initiator and the mediator according to this message:
```
03283 Raiden event	2018-12-17 15:08:17.401663	raiden.raiden_service	debug	greenlet_name=sync_handle_response-s1134736_3554614_0_1_1155_1_1_7686_1 node=82641569 raiden_event={"payment_identifier": "1545059296", "reason": "Invalid LockedTransfer message. Nonce did not change sequentially, expected: 4 got: 7.", "_type": "raiden.transfer.events.EventInvalidReceivedLockedTransfer", "_version": 0}
```

After investigating, this became obvious in these logs:
Nonce 1
```
<ActionInitMediator from_route:<RouteState hop:82641569 channel_identifier:3> from_transfer:<LockedTransferSignedState msgid:10868911650935270878 id:1545058842 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:6425 secrethash:925f9fc3> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
         <SendProcessed confirmed_msgid:10868911650935270878 recipient:82641569>
         <SendRefundTransfer msgid:8771487178460512541 transfer:<LockedTransferUnsignedState id:1545058842 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:1 transferred_amount:0 locked_amount:1 locksroot:6063fece token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:6425 secrethash:925f9fc3> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9> recipient:82641569 >
```
Nonce 2:
```
<ActionInitMediator from_route:<RouteState hop:82641569 channel_identifier:3> from_transfer:<LockedTransferSignedState msgid:9219899581963805609 id:1545058872 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:6454 secrethash:28632fa1> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
         <SendProcessed confirmed_msgid:9219899581963805609 recipient:82641569>
         <SendRefundTransfer msgid:7205785928590549623 transfer:<LockedTransferUnsignedState id:1545058872 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:2 transferred_amount:0 locked_amount:2 locksroot:44c431a5 token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:6454 secrethash:28632fa1> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9> recipient:82641569 >
```
Nonce 3:
```
<ActionInitMediator from_route:<RouteState hop:82641569 channel_identifier:3> from_transfer:<LockedTransferSignedState msgid:8908978951516977457 id:1545058903 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:6484 secrethash:3a46edfb> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
         <SendProcessed confirmed_msgid:8908978951516977457 recipient:82641569>
         <SendRefundTransfer msgid:6014380788349923296 transfer:<LockedTransferUnsignedState id:1545058903 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:3 transferred_amount:0 locked_amount:3 locksroot:cf08cb97 token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:6484 secrethash:3a46edfb> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9> recipient:82641569 >
```

Some blocks afterwards, these transfers expire so the mediator would send out `LockExpired` messages to the initiator related to all refund transfers sent previously:
Nonce 4:
```
<Block number=6435 gas_limit=8000000 block_hash=9894f941>
         <EventUnlockClaimFailed id:1545058842 secrethash:925f9fc3 reason:lock expired>
         <SendLockExpired msgid:1917831621134351790 balance_proof:<BalanceProofUnsignedState nonce:4 transferred_amount:0 locked_amount:2 locksroot:8698151e token_network:15008736 channel_identifier:3 chain_id:17> secrethash:925f9fc3 recipient:82641569>
```
Nonce 5:
```
<Block number=6464 gas_limit=8000000 block_hash=444b6a82>
         <EventUnlockClaimFailed id:1545058872 secrethash:28632fa1 reason:lock expired>
         <SendLockExpired msgid:8517134422664701001 balance_proof:<BalanceProofUnsignedState nonce:5 transferred_amount:0 locked_amount:1 locksroot:7e1548e0 token_network:15008736 channel_identifier:3 chain_id:17> secrethash:28632fa1 recipient:82641569>
```

Nonce 6:
```
<Block number=6494 gas_limit=8000000 block_hash=4219622d>
         <EventUnlockClaimFailed id:1545058903 secrethash:3a46edfb reason:lock expired>
         <SendLockExpired msgid:271834321607013906 balance_proof:<BalanceProofUnsignedState nonce:6 transferred_amount:0 locked_amount:0 locksroot:00000000 token_network:15008736 channel_identifier:3 chain_id:17> secrethash:3a46edfb recipient:82641569
```

So far, from a mediator perspective, the partner state's nonce is 6. So now it receives a locked transfer that is also refunded:
```
<ActionInitMediator from_route:<RouteState hop:82641569 channel_identifier:3> from_transfer:<LockedTransferSignedState msgid:13861462740367797137 id:1545059296 token:0x28104ee15e1c70c421150865c3fb731c426929e6 lock:<HashTimeLockState amount:1 expiration:6863 secrethash:1d59b33b> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9>>
         <SendProcessed confirmed_msgid:13861462740367797137 recipient:82641569>
         <SendRefundTransfer msgid:3803582215950841833 transfer:<LockedTransferUnsignedState id:1545059296 token:0x28104ee15e1c70c421150865c3fb731c426929e6 balance_proof:<BalanceProofUnsignedState nonce:7 transferred_amount:0 locked_amount:1 locksroot:7641a299 token_network:15008736 channel_identifier:3 chain_id:17> lock:<HashTimeLockState amount:1 expiration:6863 secrethash:1d59b33b> target:0x60ffe4dade358246c49c4a16bb5b430d333b5ce9> recipient:82641569 >
```
with nonce 7. But from the initiator's perspective, `LockExpired` isn't a state-changing message (yet) so the nonce for the initiator's `partner_state` doesn't get updated and stays at 3 (after the last refund transfer). 

What should happen is to implement a handler for the LockExpired message in the initiator so that the latest balance proof is taken from the message and recorded into the partner_state of the channel so that following messages could be processed correctly.
